src/hotspot/share/oops/oop.inline.hpp

```diff
@@ -275,14 +275,39 @@ void oopDesc::forward_to(oop p) {
   set_mark(m);
 }
 
+void oopDesc::forward_to_self() {
+  set_mark(markWord::prototype().set_self_forwarded());
+}
+
 oop oopDesc::forward_to_atomic(oop p, markWord compare, atomic_memory_order order) {
   markWord m = markWord::encode_pointer_as_mark(p);
-  assert(m.decode_pointer() == p, "encoding must be reversible");
+  assert(forwardee(m) == p, "encoding must be reversible");
   markWord old_mark = cas_set_mark(m, compare, order);
   if (old_mark == compare) {
     return nullptr;
   } else {
-    return cast_to_oop(old_mark.decode_pointer());
+    assert(old_mark.is_marked(), "must be marked here");
+    return forwardee(old_mark);
+  }
+}
+
+oop oopDesc::forward_to_self_atomic(markWord compare, atomic_memory_order order) {
+  markWord new_mark = markWord::prototype().set_self_forwarded();
+  assert(forwardee(new_mark) == cast_to_oop(this), "encoding must be reversible");
+  markWord old_mark = cas_set_mark(new_mark, compare, order);
+  if (old_mark == compare) {
+    return nullptr;
+  } else {
+    assert(old_mark.is_marked(), "must be marked here");
+    return forwardee(old_mark);
+  }
+}
+oop oopDesc::forwardee(markWord mark) const {
+  assert(mark.is_marked(), "only decode when actually forwarded");
```

Comment:
I think it would be nice if this assert could instead be written as:
```
assert(mark.is_forwarded(), "only decode when actually forwarded");
```

Would people be opposed to adding a `markWord::is_forwarded` function? We already have one for oopDesc:
```
bool oopDesc::is_forwarded() const {
  // The extra heap check is needed since the obj might be locked, in which case the
  // mark would point to a stack location and have the sentinel bit cleared
  return mark().is_marked();
}
```
(which also could use `is_forwarded` instead of `is_marked`. (BTW, the comment is old and should probably be removed) ).

Reply:
Yes, I prefer that, too. Also, once the SerialGC gets its new (compressor) GC (see #17312), we can remove the 'marked' language altogether, because the bit-pattern no longer means 'marked' but only 'forwarded'. In-fact, we could even (finally?) rename the markWord to header (or a variant like objectHeader) because it would no longer be used for marking at all.